PocketKit/Sources/Sync/Operations/RetriableOperation.swift

```diff
@@ -1,64 +1,116 @@
-        case retry
-        case success
-    }
-
     private let operation: SyncOperation
     private let retrySignal: RetrySignal
     private let backgroundTaskManager: BackgroundTaskManager
+    private let space: Space
+    private let syncTaskId: NSManagedObjectID
     private var retries = 0
     private var subscription: AnyCancellable?
 
     init(
         retrySignal: RetrySignal,
         backgroundTaskManager: BackgroundTaskManager,
-        operation: SyncOperation
+        operation: SyncOperation,
+        space: Space,
+        syncTaskId: NSManagedObjectID
     ) {
         self.retrySignal = retrySignal
         self.backgroundTaskManager = backgroundTaskManager
         self.operation = operation
+        self.space = space
+        self.syncTaskId = syncTaskId
     }
 
+    /// Performs all the main work for this operation
     override func main() {
+        /// Wrap the code in a Task so that it is Async, Apple by default does not provide an Async capable NSOperation.
+        /// But they provide a variable to make one Async.
+        /// See AsyncOperation which makes NSOperation Async
         Task {
-            let taskID = backgroundTaskManager.beginTask()
+            let taskID = backgroundTaskManager.beginTask(withName: String(describing: type(of: operation))) { [weak self] in
+                guard let self else {
+                    Log.captureNilWeakSelf()
+                    return
+                }
+                cancelOperation()
+            }
 
             switch await operation.execute() {
             case .retry(let error):
+                Log.info("Retrying persistent task with objectID \(String(describing: syncTaskId)) due to \(String(describing: error))")
                 retry(error)
             case .failure(let error):
-                Log.breadcrumb(category: "sync", level: .error, message: error.localizedDescription)
+                Log.warning("Failed persistent task with objectID \(String(describing: syncTaskId)) due to \(String(describing: error))")
                 Log.capture(error: error)
-                finishOperation()
+                doneWithTask()
             case .success:
-                finishOperation()
+                Log.info("Successfully finished persistent task with objectID \(String(describing: syncTaskId))")
+                doneWithTask()
             }
             backgroundTaskManager.endTask(taskID)
         }
     }
 
+    /// Signals that we are done with the task and will perform all clean up operations
+    private func doneWithTask() {
+        clearPersistentTask()
+        finishOperation()
+    }
+
+    /// Removes the persistent task from CoreData so that it is not executed again
+    private func clearPersistentTask() {
+        do {
+            Log.info("Deleting persistent task with objectID \(self.syncTaskId)")
+            try space.delete([self.syncTaskId], for: PersistentSyncTask.entity())
+            try space.save()
```

Comment:
nit: do you think it'd be worth  moving the `save()` call inside the `delete()` call in `Space`, just to abstract it away from here?

Reply:
talked live keeping outside so we can do things in a transaction.